src/components/TextField/TextField.js

```diff
@@ -0,0 +1,319 @@
+        }
+    }
+
+    handleInputFocus = (event) => {
+        const onFocus = this.props.onFocus || (() => {});
+        this.setState({
+            hasFocus: true,
+            showTooltip: true,
+        }, () => {
+            onFocus(event);
+        });
+    }
+
+    handleInputKeyDown = (event) => {
+        const { onKeyDown, onEnterKey } = this.props;
+        if (onKeyDown) {
+            onKeyDown(event.key, event);
+        }
+        if (onEnterKey && event.key === 'Enter') {
+            onEnterKey(event);
+        }
+    }
+
+    handleInputKeyPress = (event) => {
+        const { onKeyPress } = this.props;
+        if (onKeyPress) {
+            onKeyPress(event.key, event);
+        }
+    }
+
+    handleInputKeyUp = (event) => {
+        const { onKeyUp } = this.props;
+        if (onKeyUp) {
+            onKeyUp(event.key, event);
+        }
+    }
+
+    handleClearIconClick = () => {
+        this.setState({ value: '' });
+    }
+
+    /**
+     * wrapInputWithTooltip
+     * provide a string or component to tooltipError or tooltipHint and this
+     * wrapper will be used to create a tooltipBox (an error box is used for tooltipError)
+     * @param {Element} input - the input field
+     * @param {Component|string} tooltip - the error or hint tooltip
+     */
+    wrapInputWithTooltip = (input, tooltip) => {
+        const { DEFAULT, ERROR } = TooltipBoxStatus;
+        return (
+            tooltip ?
+                <Tooltip
+                    position={TooltipPosition.BOTTOM_CENTER}
+                    showTooltip={this.state.showTooltip}
+                    tooltip={
+                        <TooltipBox
+                            status={this.props.tooltipError ? ERROR : DEFAULT}
+                        >
+                            {tooltip}
+                        </TooltipBox>
+                    }
+                >
+                    {input}
+                </Tooltip> :
+                input
+        );
+    }
+
+    render() {
+        const showLabel = (
+            this.props.hasLabelAlways ||
+            this.state.hasFocus ||
+            this.state.hasMouseOver ||
+            !this.state.value
+        );
+        const requiredIcon = this.props.isRequired ?
+            <Tooltip tooltip={this.props.tooltipRequired}><IconRequired /></Tooltip> :
+            null;
+        /* eslint-disable jsx-a11y/label-has-for */
+        // @NB: jsx-a11y/label-has-for fails with UID as id
+        const label = this.props.label && showLabel ?
+            (
+                <label
+                    htmlFor={this.uid}
+                    className={cx(
+                        'uir-textfield-label',
+                        {
+                            'uir-textfield-label--visually-hidden': this.props.variant === TextFieldVariant.TITLE,
+                        },
+                    )}
+                >
+                    {this.props.label}
+                    {requiredIcon}
```

Comment:
Because this got moved into the `label`, we now don't show the required icon for title fields. I think that's wrong. If label is not visible, the required icon should still be rendered.